File: src/pkg04_service_cpp/src/client_node.cpp
// client_node.cpp
// Calls /add_two_ints once with values taken from command line args
// (or 3 and 5 by default), prints the response, and exits.

#include <chrono>
#include <cstdlib>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "pkg03_tutorial_interfaces/srv/add_two_ints.hpp"

using AddTwoInts = pkg03_tutorial_interfaces::srv::AddTwoInts;
using namespace std::chrono_literals;

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);

  auto node = rclcpp::Node::make_shared("add_two_ints_client");
  auto client = node->create_client<AddTwoInts>("add_two_ints");

  // Wait until the server is up (or we get tired)
  while (!client->wait_for_service(1s)) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(node->get_logger(), "Interrupted while waiting for service.");
      return 1;
    }
    RCLCPP_INFO(node->get_logger(), "Service not available yet, waiting...");
  }

  // Build the request
  auto request = std::make_shared<AddTwoInts::Request>();
  request->a = (argc > 1) ? std::atoll(argv[1]) : 3;
  request->b = (argc > 2) ? std::atoll(argv[2]) : 5;

  RCLCPP_INFO(node->get_logger(), "Sending: a=%ld, b=%ld", request->a, request->b);

  // async_send_request returns a future. We spin until it completes.
  auto future = client->async_send_request(request);

  if (rclcpp::spin_until_future_complete(node, future) ==
      rclcpp::FutureReturnCode::SUCCESS)
  {
    RCLCPP_INFO(node->get_logger(), "Result: %ld", future.get()->sum);
  } else {
    RCLCPP_ERROR(node->get_logger(), "Failed to call /add_two_ints.");
  }

  rclcpp::shutdown();
  return 0;
}

File: src/pkg05_service_py/pkg05_service_py/client_node.py
#!/usr/bin/env python3
"""Service client in Python. Calls /add_two_ints once and exits."""

import sys
import rclpy
from rclpy.node import Node
from pkg03_tutorial_interfaces.srv import AddTwoInts


class ClientNode(Node):
    def __init__(self):
        super().__init__('add_two_ints_client_py')
        self.client = self.create_client(AddTwoInts, 'add_two_ints')

        # Block until server is available
        while not self.client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('Service not available yet, waiting...')

    def send_request(self, a, b):
        request = AddTwoInts.Request()
        request.a = a
        request.b = b
        # Returns a Future. spin_until_future_complete blocks until done.
        future = self.client.call_async(request)
        rclpy.spin_until_future_complete(self, future)
        return future.result()


def main(args=None):
    rclpy.init(args=args)
    node = ClientNode()

    a = int(sys.argv[1]) if len(sys.argv) > 1 else 3
    b = int(sys.argv[2]) if len(sys.argv) > 2 else 5

    node.get_logger().info(f'Sending: a={a}, b={b}')
    response = node.send_request(a, b)
    node.get_logger().info(f'Result: {response.sum}')

    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
